### ISSUE TITLE
Impure expressions allowed in macro splices

### ISSUE THREAD

**Issue by yallop** [modular-macros/ocaml-macros#8]
The implementation currently rejects the following program

```ocaml
macro x = (~Pervasives.print_endline "hello, world"; << () >>)
```
with the following error message:
```
Error: Impure constructions are not allowed as macro definitions
```

However, quotations with side-effects in splices are accepted as macro bodies.  For example, the following program is accepted

```ocaml
macro x = << $(~Pervasives.print_endline "hello, world";  <<()>>) >>
let _ = ($(x), $(x))
```

and prints

```
hello, world
hello, world
```

**Comment by lpw25** [comment]
Well spotted. [This](https://github.com/ocamllabs/ocaml-macros/blob/macros_with_lambda_quotes/typing/typecore.ml#L1633) is the line at fault. It needs to recurse into the quotation and check any splices for non-expansiveness/purity.

**Comment by yallop** [comment]
Fixed by #26.
